archive/imo/imo2008_q2.lean

```diff
@@ -0,0 +1,156 @@
+
+  calc  x^2 / (x-1)^2 + y^2 / (y-1)^2 + z^2 / (z-1)^2 - 1
+      = (a/b)^2 / (a/b-1)^2 + (b/c)^2 / (b/c-1)^2 + (c/a)^2 / (c/a-1)^2 - 1 :
+        by rw [hx₂, hy₂, hz₂]
+  ... = a^2/(a-b)^2 + b^2/(b-c)^2 + c^2/(c-a)^2 - 1 :
+        by field_simp [div_sub_one hb, div_sub_one hc, div_sub_one ha]
+  ... = (c-m-n)^2/n^2 + (c-m)^2/m^2 + c^2/(m+n)^2 - 1 :
+        by rw [hab_mn, hbc_mn, hca_mn, ha_cmn, hb_cmn]
+  ... = ( (c*(m^2+n^2+m*n) - m*(m+n)^2) / (m*n*(m+n)) )^2 :
+        by { ring_nf, field_simp, ring }
+  ... ≥ 0 :
+        by exact pow_two_nonneg _ ,
+
+  linarith [key],
+end
+
+def rational_solutions := { s : ℚ×ℚ×ℚ | ∃ (x y z : ℚ), s = (x, y, z) ∧
+  x ≠ 1 ∧ y ≠ 1 ∧ z ≠ 1 ∧ x*y*z = 1 ∧ x^2/(x-1)^2 + y^2/(y-1)^2 + z^2/(z-1)^2 = 1 }
+
+theorem imo2008_q2b : set.infinite rational_solutions :=
+begin
+  let W := { s : ℚ×ℚ×ℚ | ∃ (x y z : ℚ), s = (x, y, z) ∧
+    ∃ t : ℚ, t > 0 ∧ x = -(t+1)/t^2 ∧ y = t/(t+1)^2 ∧ z = -t*(t+1)},
+
+  have hW_sub_S : W ⊆ rational_solutions,
+  { intros s hs_in_W,
+    rw rational_solutions,
+    simp at hs_in_W ⊢,
+    rcases hs_in_W with ⟨x, y, z, h₁, ht⟩,
+    rcases ht with ⟨t, ht_gt_zero, hx_t, hy_t, hz_t⟩,
+    use [x, y, z],
+
+    have ht_ne_zero  : t ≠ 0,           exact ne_of_gt ht_gt_zero,
+    have ht1_ne_zero : t+1 ≠ 0,         linarith[ht_gt_zero],
+    have key_gt_zero : t^2 + t + 1 > 0, linarith[pow_pos ht_gt_zero 2, ht_gt_zero],
+    have key_ne_zero : t^2 + t + 1 ≠ 0, exact ne_of_gt key_gt_zero,
+
+    have h₂ : x ≠ 1,      { rw hx_t, field_simp, linarith[key_gt_zero] },
+    have h₃ : y ≠ 1,      { rw hy_t, field_simp, linarith[key_gt_zero] },
+    have h₄ : z ≠ 1,      { rw hz_t, linarith[key_gt_zero] },
+    have h₅ : x*y*z = 1,  { rw [hx_t, hy_t, hz_t], field_simp, ring },
+
+    have h₆ : x^2/(x-1)^2 + y^2/(y-1)^2 + z^2/(z-1)^2 = 1,
+    { have hx1 : (x - 1)^2 = (t^2 + t + 1)^2/t^4,     { field_simp, rw hx_t, field_simp, ring },
+      have hy1 : (y - 1)^2 = (t^2 + t + 1)^2/(t+1)^4, { field_simp, rw hy_t, field_simp, ring },
+      have hz1 : (z - 1)^2 = (t^2 + t + 1)^2,         { rw hz_t, ring },
+      calc  x^2/(x-1)^2 + y^2/(y-1)^2 + z^2/(z-1)^2
+          = (x^2*t^4 + y^2*(t+1)^4 + z^2)/(t^2 + t + 1)^2 :
+            by { rw [hx1, hy1, hz1], field_simp }
+      ... = 1 :
+            by { rw [hx_t, hy_t, hz_t], field_simp, ring } },
+
+    exact ⟨h₁, h₂, h₃, h₄, h₅, h₆⟩ },
+
+  have hW_inf : set.infinite W,
+  { set g : ℚ×ℚ×ℚ → ℚ := (λs, -s.2.2) with hg_eval,
+    let K := g '' W,
+
+    have hK_not_bdd : ¬bdd_above K,
+    { rw not_bdd_above_iff,
+      intro q,
+      let t : ℚ := max (q+1) 1,
+      use t*(t+1),
+
+      have h₁ : t * (t + 1) ∈ K,
+      { let x : ℚ := (-1 + -t)/t^2,
+        let y : ℚ := t/(t+1)^2,
+        set z : ℚ := -t*(t+1) with hz_def,
+
+        simp, use [x, y, z], split,
+        { use [x, y, z], split,
+          exact ⟨refl x, refl y, refl z⟩,
+          { use t, split,
+            { simp, right, exact zero_lt_one },
```

Comment:
This is a "non-terminal" `simp`, which might break if we add more `simp` lemmas in the future. Can you try `simp?` or `squeeze_simp` here and replace this with the `simp only ...` output?

Reply:
done!